app/scripts/controllers/metamask-notifications/metamask-notifications.ts

```diff
@@ -0,0 +1,955 @@
+ * This controller subscribes to account state changes and ensures
+ * that the account list is updated based on the latest account configurations.
+ */
+export class MetamaskNotificationsController extends BaseController<
+  typeof controllerName,
+  MetamaskNotificationsControllerState,
+  MetamaskNotificationsControllerMessenger
+> {
+  #auth = {
+    getBearerToken: async () => {
+      return await this.messagingSystem.call(
+        'AuthenticationController:getBearerToken',
+      );
+    },
+    isSignedIn: () => {
+      return this.messagingSystem.call('AuthenticationController:isSignedIn');
+    },
+  };
+
+  #storage = {
+    getStorageKey: () => {
+      return this.messagingSystem.call('UserStorageController:getStorageKey');
+    },
+    getNotificationStorage: async () => {
+      return await this.messagingSystem.call(
+        'UserStorageController:performGetStorage',
+        'notification_settings',
+      );
+    },
+    setNotificationStorage: async (state: string) => {
+      return await this.messagingSystem.call(
+        'UserStorageController:performSetStorage',
+        'notification_settings',
+        state,
+      );
+    },
+  };
+
+  #pushNotifications = {
+    enablePushNotifications: async (UUIDs: string[]) => {
+      return await this.messagingSystem.call(
+        'PushPlatformNotificationsController:enablePushNotifications',
+        UUIDs,
+      );
+    },
+    disablePushNotifications: async (UUIDs: string[]) => {
+      return await this.messagingSystem.call(
+        'PushPlatformNotificationsController:disablePushNotifications',
+        UUIDs,
+      );
+    },
+    updatePushNotifications: async (UUIDs: string[]) => {
+      return await this.messagingSystem.call(
+        'PushPlatformNotificationsController:updateTriggerPushNotifications',
+        UUIDs,
+      );
+    },
+  };
+
+  #accountsCache = new Set<string>();
+
+  #accounts = {
+    listAccounts: () => {
+      // Fetch the list of accounts
+      const accounts = this.messagingSystem.call(
+        'AccountsController:listAccounts',
+      ) as InternalAccount[];
+
+      // Extract addresses and convert them to checksum addresses to ensure case sensitivity is not an issue
+      const addresses = accounts.map((account) => {
+        return toChecksumHexAddress(account.address as string);
+      });
+
+      this.#accountsCache = new Set([...this.#accountsCache, ...addresses]);
+      return [...this.#accountsCache];
+    },
+
+    // This ensures that our cache is up-to-date when constructor is called.
+    initialize: () => {
+      this.#accounts.listAccounts();
+    },
+
+    // Subscription when an new account is changed
+    // The initial cache should already have addresses already have in wallet,
+    // so any new account switched to (that is not in cache) We can assume it is new.
+    createNewNotificationsOnNewAccountChange: () => {
+      this.messagingSystem.subscribe(
+        'AccountsController:selectedAccountChange',
```

Comment:
As the comment suggested - we had to use this event for figuring out if new addresses were added to the wallet.

Is there are more reliable event we can listen to?

Reply:
Is there is a concern about reliability here? Does `AccountsController:selectedAccountChange` not occur when you would expect it to?

Reply:
I wonder how it behaves for adding multiple accounts during hardware wallet connection flows :thinking:

Reply:
Also, we may want to handle the case where hardware wallet accounts, or snap accounts, are removed.

Reply:
The `AccountsController:selectedAccountChange` does work as we intended (so far) - but was just wondering if there is a specific event just for adding new accounts instead of when an account is changed.

I'll do some additional testing to ensure we are targeting the correct accounts - great callout!

Reply:
> accounts removed

Great callout! Will ned to check this.

Reply:
If this is a bit more challenging than expected then we can have this in another PR

Reply:
Instead of using the AccountsController, we've opted for the Keyring Controller does a pretty good job for our use-case!

Also covered multi-add or multi-remove of accounts.